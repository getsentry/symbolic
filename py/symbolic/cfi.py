"""Minidump processing"""

import shutil

from symbolic._lowlevel import lib, ffi
from symbolic.utils import (
    RustObject,
    rustcall,
    encode_path,
    make_buffered_slice_reader,
)

__all__ = [
    "CfiCache",
    "CFICACHE_LATEST_VERSION",
]

# The most recent version for the CFI cache file format
CFICACHE_LATEST_VERSION = rustcall(lib.symbolic_cficache_latest_version)


class CfiCache(RustObject):
    """A cache for call frame information (CFI) to improve native stackwalking"""

    __dealloc_func__ = lib.symbolic_cficache_free

    @classmethod
    def open(cls, path):
        """Loads a cficache from a file via mmap."""
        return cls._from_objptr(rustcall(lib.symbolic_cficache_open, encode_path(path)))

    @classmethod
    def from_object(cls, obj):
        """Creates a cficache from the given object."""
        return cls._from_objptr(
            rustcall(lib.symbolic_cficache_from_object, obj._get_objptr())
        )

    @property
    def version(self):
        """Version of the file format."""
        return self._methodcall(lib.symbolic_cficache_get_version)

    @property
    def is_latest_version(self):
        """Returns true if this is the latest file format."""
        return self.version >= CFICACHE_LATEST_VERSION

    def open_stream(self):
        """Returns a stream to read files from the internal buffer."""
        buf = self._methodcall(lib.symbolic_cficache_get_bytes)
        size = self._methodcall(lib.symbolic_cficache_get_size)
        return make_buffered_slice_reader(ffi.buffer(buf, size), self)

    def write_to(self, f):
        """Writes the CFI cache into a file object."""
        shutil.copyfileobj(self.open_stream(), f)
