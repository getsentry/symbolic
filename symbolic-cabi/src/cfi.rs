use std::ffi::CStr;
use std::os::raw::c_char;

use symbolic::cfi::{CfiCache, CFICACHE_LATEST_VERSION};
use symbolic::common::ByteView;

use crate::debuginfo::SymbolicObject;
use crate::utils::ForeignObject;

/// Contains stack frame information (CFI) for an image.
pub struct SymbolicCfiCache;

impl ForeignObject for SymbolicCfiCache {
    type RustObject = CfiCache<'static>;
}

ffi_fn! {
    /// Extracts call frame information (CFI) from an Object.
    unsafe fn symbolic_cficache_from_object(
        object: *const SymbolicObject,
    ) -> Result<*mut SymbolicCfiCache> {
        let object = SymbolicObject::as_rust(object).get();
        let cache = CfiCache::from_object(object)?;
        Ok(SymbolicCfiCache::from_rust(cache))
    }
}

ffi_fn! {
    /// Loads a CFI cache from the given path.
    unsafe fn symbolic_cficache_open(path: *const c_char) -> Result<*mut SymbolicCfiCache> {
        let byteview = ByteView::open(CStr::from_ptr(path).to_str()?)?;
        let cache = CfiCache::from_bytes(byteview)?;
        Ok(SymbolicCfiCache::from_rust(cache))
    }
}

ffi_fn! {
    /// Returns the file format version of the CFI cache.
    unsafe fn symbolic_cficache_get_version(cache: *const SymbolicCfiCache) -> Result<u32> {
        Ok(SymbolicCfiCache::as_rust(cache).version())
    }
}

ffi_fn! {
    /// Returns a pointer to the raw buffer of the CFI cache.
    unsafe fn symbolic_cficache_get_bytes(cache: *const SymbolicCfiCache) -> Result<*const u8> {
        Ok(SymbolicCfiCache::as_rust(cache).as_slice().as_ptr())
    }
}

ffi_fn! {
    /// Returns the size of the raw buffer of the CFI cache.
    unsafe fn symbolic_cficache_get_size(cache: *const SymbolicCfiCache) -> Result<usize> {
        Ok(SymbolicCfiCache::as_rust(cache).as_slice().len())
    }
}

ffi_fn! {
    /// Releases memory held by an unmanaged `SymbolicCfiCache` instance.
    unsafe fn symbolic_cficache_free(cache: *mut SymbolicCfiCache) {
        SymbolicCfiCache::drop(cache);
    }
}

ffi_fn! {
    /// Returns the latest CFI cache version.
    unsafe fn symbolic_cficache_latest_version() -> Result<u32> {
        Ok(CFICACHE_LATEST_VERSION)
    }
}
