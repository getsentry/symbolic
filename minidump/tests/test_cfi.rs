use std::str;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::Object;
use symbolic_minidump::cfi::{AsciiCfiWriter, CfiCache};

#[test]
fn load_empty_cfi_cache() -> Result<(), Error> {
    let buffer = ByteView::from_slice(&[]);
    let cache = CfiCache::from_bytes(buffer)?;
    assert_eq!(cache.version(), 1);
    Ok(())
}

#[test]
fn cfi_from_elf() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    // NOTE: Breakpad's CFI writer outputs registers in alphabetical order. We
    // write the CFA register first, and then order by register number. Thus,
    // the output is not identical to `cfi_sym_linux`.
    insta::assert_snapshot_matches!("cfi_elf", cfi);

    Ok(())
}

#[test]
fn cfi_from_macho() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    // NOTE: Breakpad's CFI writer outputs registers in alphabetical order. We
    // write the CFA register first, and then order by register number. Thus,
    // the output is not identical to `cfi_sym_macos`.
    insta::assert_snapshot_matches!("cfi_macho", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/crash.sym")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot_matches!("cfi_sym_linux", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/crash.sym")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot_matches!("cfi_sym_macos", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot_matches!("cfi_sym_windows", cfi);

    Ok(())
}

#[test]
fn cfi_from_pdb_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot_matches!("cfi_pdb_windows", cfi);

    Ok(())
}

#[test]
fn cfi_from_pe_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot_matches!("cfi_pe_windows", cfi);

    Ok(())
}
