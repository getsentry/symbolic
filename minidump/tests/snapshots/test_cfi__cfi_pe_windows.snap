---
created: "2019-04-03T18:49:08.328139Z"
creator: insta@0.7.4
source: minidump/tests/test_cfi.rs
expression: cfi
---
STACK CFI INIT 1010 55 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1010 .cfa: $rsp 80 +
STACK CFI INIT 10b0 43 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 10b0 .cfa: $rsp 48 +
STACK CFI INIT 1120 2d .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1120 .cfa: $rsp 48 +
STACK CFI INIT 1150 18 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1150 .cfa: $rsp 64 +
STACK CFI INIT 1170 a .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1170 .cfa: $rsp 48 +
STACK CFI INIT 1190 21 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1190 .cfa: $rsp 8 +
STACK CFI INIT 11bc 3c .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 11bc .cfa: $rsp 48 +
STACK CFI INIT 11f8 2b .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 11f8 .cfa: $rsp 48 +
STACK CFI INIT 1224 b7 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1224 .cfa: $rsp 48 +
STACK CFI INIT 12dc 10 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 12dc .cfa: $rsp 48 +
STACK CFI INIT 12ec 19 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 12ec .cfa: $rsp 48 +
STACK CFI INIT 1308 17c .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1308 .cfa: $rsp 64 +
STACK CFI INIT 1484 12 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1484 .cfa: $rsp 48 +
STACK CFI INIT 1498 34 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1498 .cfa: $rsp 48 +
STACK CFI INIT 14cc d1 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 14cc .cfa: $rsp 64 +
STACK CFI INIT 15a0 71 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 15a0 .cfa: $rsp 96 +
STACK CFI INIT 161c 3f .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 161c .cfa: $rsp 48 +
STACK CFI INIT 167c 3f .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 167c .cfa: $rsp 48 +
STACK CFI INIT 16dc 35 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 16dc .cfa: $rsp 48 +
STACK CFI INIT 1728 42 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1728 .cfa: $rsp 48 +
STACK CFI INIT 176c 20 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 176c .cfa: $rsp 80 +
STACK CFI INIT 178c 20 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 178c .cfa: $rsp 80 +
STACK CFI INIT 17ac 39 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 17ac .cfa: $rsp 48 +
STACK CFI INIT 17e8 49 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 17e8 .cfa: $rsp 48 +
STACK CFI INIT 1834 d8 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1834 .cfa: $rsp 80 +
STACK CFI INIT 190c 99 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 190c .cfa: $rsp 32 +
STACK CFI INIT 19a8 24 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 19a8 .cfa: $rsp 48 +
STACK CFI INIT 19cc 2b .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 19cc .cfa: $rsp 48 +
STACK CFI INIT 19f8 4f .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 19f8 .cfa: $rsp 48 +
STACK CFI INIT 1a48 17 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1a48 .cfa: $rsp 48 +
STACK CFI INIT 1a60 ac .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1a60 .cfa: $rsp 48 +
STACK CFI INIT 1b40 1b .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1b40 .cfa: $rsp 48 +
STACK CFI INIT 1b80 14a .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1b80 .cfa: $rsp 1488 +
STACK CFI INIT 1cd4 52 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1cd4 .cfa: $rsp 48 +
STACK CFI INIT 1d38 38 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1d38 .cfa: $rsp 48 +
STACK CFI INIT 1d70 3c .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1d70 .cfa: $rsp 48 +
STACK CFI INIT 1dac 3c .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1dac .cfa: $rsp 48 +
STACK CFI INIT 1de8 1b9 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 1de8 .cfa: $rsp 48 +
STACK CFI INIT 2090 2 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 2090 .cfa: $rsp 8 +
STACK CFI INIT 20ac 1e .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 20ac .cfa: $rsp 48 +
STACK CFI INIT 20ca 18 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI 20ca .cfa: $rsp 16 +

